core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala

```diff
@@ -509,20 +517,57 @@ class SpendReportingService(
         )
       }
 
-      query = getQuery(aggregations, spendExportConf)
-      queryJob = setUpQuery(query, spendExportConf, start, end, projectNames)
+      // look for a cached spend report
+      isCacheable = aggregationsAreCacheable(aggregations)
+      cachedResults <-
+        if (isCacheable)
+          getCachedSpendReportData(projectNames, start, end)
+        else
+          Future.successful(CachedSpendReportingResults(isCacheValid = false, results = None))
+
+      spendResults <-
+        if (cachedResults.isCacheValid) {
+          rawlsCacheHitRate().hit()
+          cachedResults.results match {
+            case Some(report) => Future.successful(report)
+            case None =>
+              Future.failed(
+                RawlsExceptionWithErrorReport(
+                  StatusCodes.NotFound,
+                  s"no spend data found for billing project ${project.value} between dates ${toISODateString(start)} and ${toISODateString(end)}"
+                )
+              )
+          }
+        } else {
+          // TODO CORE-291: should this log a miss even for Daily aggregations?
+          rawlsCacheHitRate().miss()
+          val query = getQuery(aggregations, spendExportConf)
+          val queryJob = setUpQuery(query, spendExportConf, start, end, projectNames)
+
+          val queryResults = runBigQueryJob(queryJob, childContext).map { result =>
+            result.getValues.asScala.toList match {
+              case Nil  => None
+              case rows => Option(extractSpendReportingResults(rows, start, end, projectNames, aggregations))
+            }
+          }
+
+          // write BigQuery results back to cache
+          if (isCacheable) {
+            queryResults.map { res =>
+              insertRecordsWithMissingSpendData(res, projectNames, start, end)
```

Comment:
taking a closer look at this, `insertRecordsWithMissingSpendData` returns `Set[Future[...]]` which is an odd signature. `Future[Set[..]]` would be more expected. But I also note that these Future are dropped. Maybe that is intentional.

Reply:
I agree the return value is unexpected … also agree it doesn't matter too much because this spend report and the consolidated spend report both fire-and-forget this. It feels ok to return the result to the user without waiting for these inserts.

Reply:
Apologies for the Scala strangeness 😅